### PR TITLE
File Additions (#23 #24 #25)

### PR DESCRIPTION
**Pull request by OctopusProteins**
- Added /timelapse #24 
 - Added /getfile #23 
 - Added /gettimelapse #25 
 - Added new setting option base URL

### CODE REVIEW

File: octoprint_discordremote/command.py [hunk 1]
```diff
@@ -45,6 +53,90 @@ def parse_command(self, string):
...
+            if not response:
+                return None, error_embed(title="ConnectionError", description="URL: " + str(url))
+        except ConnectionError:
+            return None, error_embed(title="ConnectionError", description="URL: " + str(url))
+        data = response.json()
+        if response.status_code == 200:
+            return None, info_embed(title=params[2], description=(data['refs'])['download'])
+        return None, error_embed(title="File Not Found", description=params[2])
+
+    def timelapse(self):
```
Comment (cameroncros) [inline octoprint_discordremote/command.py]: @OctopusProteins Seems that the /gettimelapse and /timelapse functionality cross over a bit, i suggest merging them together. Just get rid of /gettimelapse, and add the URL into the /timelapse function?

File: octoprint_discordremote/command.py [hunk 2]
```diff
@@ -89,7 +181,7 @@ def list_files(self):
             description = ''
             title = ''
             try:
-                title = details['path']
+                title = (details['path'])[1:]
```
Comment (cameroncros) [inline octoprint_discordremote/command.py]: @OctopusProteins Would prefer details['path'][1:], no need for parenthesis

File: octoprint_discordremote/command.py [hunk 3]
```diff
@@ -25,6 +28,11 @@ def __init__(self, plugin):
         self.command_dict['/help'] = {'cmd': self.help, 'description': "Print this help."}
         self.command_dict['/pause'] = {'cmd': self.pause, 'description': "Pause current print."}
         self.command_dict['/resume'] = {'cmd': self.resume, 'description': "Resume current print."}
+        self.command_dict['/timelapse'] = {'cmd': self.timelapse, 'description': "List all timelapses."}
+        self.command_dict['/gettimelapse'] = {'cmd': self.get_timelapse, 'params': "[filename]",
+                                              'description': "Gets the download link for the specified timelapse."}
+        self.command_dict['/getfile'] = {'cmd': self.get_file, 'params': "[location] [filename]",
```
Comment (cameroncros) [inline octoprint_discordremote/command.py]: @OctopusProteins Manditory params should be written like: {location} {filename}. Square brackets are for optional params.

File: octoprint_discordremote/command.py [hunk 4]
```diff
@@ -45,6 +53,90 @@ def parse_command(self, string):
...
+                                                 "\nbase_url: " + str(shared_vars.base_url))
+
+        api_key = self.plugin.get_settings().global_get(["api", "key"])
+        port = self.plugin.get_settings().global_get(["server", "port"])
+        header = {'X-Api-Key': api_key}
+
+        response = requests.get("http://127.0.0.1:%s/api/timelapse" % port, headers=header)
+        data = response.json()
+
+        for x in data['files']:
```
[the remark this answers is not in the document]
Comment (OctopusProteins) [reply octoprint_discordremote/command.py]: Fixed by 8cc3482

File: octoprint_discordremote/command.py (hunk 2 above)
Comment (OctopusProteins) [reply octoprint_discordremote/command.py]: Fixed by 5e83f84

File: octoprint_discordremote/command.py [hunk 5]
```diff
@@ -45,6 +53,90 @@ def parse_command(self, string):
         else:
             return command['cmd']()
 
+    def get_timelapse(self, params):
+        if len(params) > 2:
+            return None, error_embed(title='Too many parameters',
+                                     description='Should be: /gettimelapse [filename]')
+        if shared_vars.base_url is None or shared_vars.base_url == "":
```
[the remark this answers is not in the document]
Comment (OctopusProteins) [reply octoprint_discordremote/command.py]: Correct me if I'm wrong, but I thought that 127.0.0.1 would only work if it is accessed by the Pi? The user wouldn't be able to use this address on their machine, which is the only case that base_url is going to be used at the moment.

File: octoprint_discordremote/command.py (hunk 3 above)
Comment (OctopusProteins) [reply octoprint_discordremote/command.py]: Fixed by 3991180
Also added error case if missing parameters

File: octoprint_discordremote/command.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cameroncros) [reply octoprint_discordremote/command.py]: hmm, good point. Could use the get_ip_address() function to get the local network IP address.

[the remark this answers is not in the document]
Comment (OctopusProteins) [reply octoprint_discordremote/command.py]: I can have it default to the local ip address, but if the user wants to be able to download the files from anywhere they have to change the setting. For example, I have my Pi behind a domain, and I want it to use that so I can access it outside of my home network.

[the remark this answers is not in the document]
Comment (OctopusProteins) [reply octoprint_discordremote/command.py]: Base URL defaults to local IP in 104a815

**Comment by cameroncros** [comment]
Im still not keen on the shared_vars class, it feels like it is duplicating the get_settings() function. The command class. There is no class that does not have access to that function via the plugin, so I think it would be best to use that. I'm still happy to merge it as is, but I will likely refactor it out at some point in the future anyway.

**Comment by OctopusProteins** [comment]
I can make some modifications to get rid of it. The problem was that I was getting stuck in import loops with command.py and __init__.py. How can I use the get_settings() function?

File: octoprint_discordremote/__init__.py [hunk 6]
```diff
@@ -144,6 +146,7 @@ def get_settings_defaults(self):
             'bottoken': "",
             'channelid': "",
             'allowedusers': "",
+            'baseurl': self.get_ip_address(),
```
Comment (cameroncros) [inline octoprint_discordremote/__init__.py]: @OctopusProteins I think default value should be None, and the logic that uses baseurl should just swap None/"" with self.get_ip_address()

File: octoprint_discordremote/command.py [hunk 7]
```diff
@@ -25,6 +28,9 @@ def __init__(self, plugin):
         self.command_dict['/help'] = {'cmd': self.help, 'description': "Print this help."}
         self.command_dict['/pause'] = {'cmd': self.pause, 'description': "Pause current print."}
         self.command_dict['/resume'] = {'cmd': self.resume, 'description': "Resume current print."}
+        self.command_dict['/timelapse'] = {'cmd': self.timelapse, 'description': "List all timelapses and download links."}
+        self.command_dict['/getfile'] = {'cmd': self.get_file, 'params': "{location} {filename}",
```
Comment (cameroncros) [inline octoprint_discordremote/command.py]: @OctopusProteins And now that I think about it, we could merge /files and /getfile fairly easily as well, and make them a bit more consistent. I am happy to merge without this change though, and I can do it?

File: octoprint_discordremote/shared_vars.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#Used for accessing variables across classes when imports are not viable
```
Comment (cameroncros) [inline octoprint_discordremote/shared_vars.py]: @OctopusProteins Im pretty sure there is not need for this. The plugin (__init__.py) class is passed into every sub class, so anything that needs to be centralised should go in the plugin, or use get_settings().

File: octoprint_discordremote/command.py (hunk 7 above)
Comment (OctopusProteins) [reply octoprint_discordremote/command.py]: Sure, go ahead ;)

File: octoprint_discordremote/__init__.py (hunk 6 above)
Comment (OctopusProteins) [reply octoprint_discordremote/__init__.py]: I can't access the ```get_ip_address()``` function from the Command class, unless I'm missing something...?

File: octoprint_discordremote/shared_vars.py (hunk 8 above)
Comment (OctopusProteins) [reply octoprint_discordremote/shared_vars.py]: On it.

File: octoprint_discordremote/__init__.py (hunk 6 above)
Comment (OctopusProteins) [reply octoprint_discordremote/__init__.py]: Fixed by 1007480

File: octoprint_discordremote/shared_vars.py (hunk 8 above)
Comment (OctopusProteins) [reply octoprint_discordremote/shared_vars.py]: Fixed by ea06074

**Comment by cameroncros** [comment]
There are some minor issues, so I will fix them up and merge it. Thanks for your help.

**Pull request merged**